giskard/ml_worker/websocket/listener.py

```diff
@@ -684,13 +686,106 @@ def echo(params: websocket.EchoMsg, *args, **kwargs) -> websocket.EchoMsg:
+    client: Optional[GiskardClient],
+    params: websocket.GetPushParam,
+    push: Optional[Push],
+    push_kind: PushKind,
+    cta_kind: CallToActionKind,
+):
+    if push is None:
+        push = get_push_objects(client, params)
+
+    logger.info("Handling push kind: %s with cta kind: %s", str(push_kind), str(cta_kind))
+
+    object_uuid = ""
+    object_params = {}
+
+    project_key = params.model.project_key
+
+    # Upload related object depending on CTA type
+    if cta_kind == CallToActionKind.CREATE_SLICE or cta_kind == CallToActionKind.CREATE_SLICE_OPEN_DEBUGGER:
+        push.slicing_function.meta.tags.append("generated")
+        object_uuid = push.slicing_function.upload(client, project_key)
+    elif cta_kind == CallToActionKind.SAVE_PERTURBATION:
+        for perturbation in push.transformation_function:
+            object_uuid = perturbation.upload(client, project_key)
+    elif cta_kind == CallToActionKind.SAVE_EXAMPLE:
+        object_uuid = push.saved_example.upload(client, project_key)
+    elif cta_kind == CallToActionKind.CREATE_TEST or cta_kind == CallToActionKind.ADD_TEST_TO_CATALOG:
+        object_params = {}
+        for test in push.tests:
+            object_uuid = test.upload(client, project_key)
+        for test_param_name, test_param in push.test_params.items():
+            if isinstance(test_param, (RegistryArtifact, Dataset)):
+                object_params[test_param_name] = test_param.upload(client, project_key)
+            else:
+                object_params[test_param_name] = test_param
+
+    if object_uuid:
+        logger.info("Uploaded object for CTA with uuid: %s", object_uuid)
+        return websocket.PushAction(object_uuid=object_uuid, arguments=function_argument_to_ws(object_params))
+
+
 @websocket_actor(MLWorkerAction.getPush, timeout=30, ignore_timeout=True)
 def get_push(
     client: Optional[GiskardClient], params: websocket.GetPushParam, *args, **kwargs
 ) -> websocket.GetPushResponse:
-    object_uuid = ""
-    object_params = {}
-    project_key = params.model.project_key
+    # Save cta_kind and push_kind and remove it from params
+    cta_kind = params.cta_kind
+    push_kind = params.push_kind
+    params.cta_kind = None
+    params.push_kind = None
+
+    kinds = (
+        [push_kind]
+        if push_kind is not None
+        else [
+            PushKind.CONTRIBUTION,
+            PushKind.OVERCONFIDENCE,
+            PushKind.BORDERLINE,
+            PushKind.PERTURBATION,
+        ]
+    )
+    all_ws_res = {}
+    push = None
+    for kind in kinds:
+        params.push_kind = kind
+        logger.info("Getting push for %s", kind)
+
+        # We get a JSON for stability across process
+        json_params = params.json()
+        _, res = CACHE.get_result("push" + json_params)
+        cache_hit_ws, res_ws = CACHE.get_result("pushws" + json_params)
+        if not cache_hit_ws:
+            res = get_push_objects(client, params)
+            res_ws = push_to_ws(res)
```

Comment:
is it that expensive to convert pushes to DTOs?

Reply:
No, I can remove it. Basically, we could not cache push objects before, so